staging/src/k8s.io/apiserver/plugin/pkg/audit/dynamic/dynamic.go

```diff
@@ -0,0 +1,292 @@
+// shared informer. It takes the static policy and static backends configured by the runtime flags
+// as arguments and runs them with policy enforcment.
+func NewEnforcedBackend(c *Config) (audit.EnforcedBackend, error) {
+	informer := c.F.Auditregistration().V1alpha1().AuditSinks()
+
+	if c.BufferedConfig == nil {
+		defaultConfig := bufferedplugin.NewDefaultBatchConfig()
+		c.BufferedConfig = &defaultConfig
+	}
+
+	backends := make(map[types.UID]*Backend)
+	var staticBackend *Backend
+	if c.StaticBackend != nil {
+		enforced := legacyenforcedplugin.NewBackend(c.StaticBackend, c.StaticPolicyChecker)
+		staticBackend = &Backend{nil, enforced, make(chan struct{})}
+		backends[staticUID] = staticBackend
+	}
+
+	cm, err := webhook.NewClientManager(auditinternal.SchemeGroupVersion, auditinternal.AddToScheme)
+	if err != nil {
+		return nil, err
+	}
+	authInfoResolver, err := webhook.NewDefaultAuthenticationInfoResolver("")
+	if err != nil {
+		return nil, err
+	}
+	cm.SetAuthenticationInfoResolver(authInfoResolver)
+	cm.SetServiceResolver(webhook.NewDefaultServiceResolver())
+	cm.SetAuthenticationInfoResolverWrapper(c.AuthInfoResolverWrapper)
+
+	manager := &dynamic{
+		config:               c,
+		staticBackend:        staticBackend,
+		backends:             syncedBackends{idMap: backends},
+		updater:              updater{},
+		lister:               informer.Lister(),
+		hasSynced:            informer.Informer().HasSynced,
+		webhookClientManager: cm,
+	}
+
+	// On any change, rebuild the config
+	informer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    func(_ interface{}) { manager.updateConfiguration() },
+		UpdateFunc: func(_, _ interface{}) { manager.updateConfiguration() },
+		DeleteFunc: func(_ interface{}) { manager.updateConfiguration() },
+	})
+
+	return manager, nil
+}
+
+type dynamic struct {
+	config               *Config
+	staticBackend        *Backend
+	backends             syncedBackends
+	updater              updater
+	lister               auditlisters.AuditSinkLister
+	hasSynced            func() bool
+	webhookClientManager webhook.ClientManager
+}
+
+type syncedBackends struct {
+	idMap map[types.UID]*Backend
+	sync.RWMutex
+}
+
+// updater forces configuration updates to run serially. This is a separate lock
+// so that events are not blocked while the backends build
+type updater struct{ sync.Mutex }
+
+// ProcessEnforcedEvents proccesses the given events per current backend map
+func (d *dynamic) ProcessEnforcedEvents(events ...*audit.EnforcedEvent) {
+	for _, b := range d.GetBackends() {
+		b.enforcedBackend.ProcessEnforcedEvents(events...)
+	}
+}
+
+// Run the dynamic backends
+func (d *dynamic) Run(stopCh <-chan struct{}) error {
+	// Shutdown channels are handled individually with the dynamic
+	// backend, message must be propagated out
+	go func() {
+		for {
+			select {
+			case <-stopCh:
+				d.Shutdown()
+				return
+			default:
+			}
```

Comment:
This is a hot loop.

Reply:
What are you trying to do here? I think you can just delete this whole goroutine. Run should not be responsible for calling shutdown.

Reply:
I see. Signalling stopCh & shutdown are separate operations. I think you can sort of interpret it as stop==start shutting down, and shutdown==wait for async operations to finish, and clean up (they could probably be named better). Given that, I would do the following:

```
go func() {
  <- stopCh
  d.disableUpdates() // prevent new backends from being added
  d.stopAllBackends()
}()
```
`stopAllBackends()` should do what `Shutdown` currently does.
`Shutdown` should just call `Shutdown()` on every backend (synchronously)